### DISCUSSION TITLE
OpenTCS Integration Example

### DISCUSSION THREAD

**Discussion by MinaTofiles** [openTCS/opentcs#78] [Q&A]
Hello,
I am a beginner in this field, so excuse my knowledge.
I am trying to integrate OpenTCS with my project, and as I've found in the docs that integration example project is the best way to go.
I've downloaded it and builded the source code and everything is running smoothly thanks to the developers.
But when i enable the vehicle driver which is Custom Comm Adapter it is enabled correctly but doesn't connect to the vehicle and state is UNKNOWN.
After debugging i get connection refused to <vehicleHost:vehiclePort>.
So, should I do anything to make it connect or this is something wrong with my setup? 
Thanks in advance.

**Comment by swltr** [comment]
That vehicle driver is part of the example integration project to as an example how to implement a vehicle driver. You need a real vehicle implementing the same communication protocol as the driver to connect to when you use this driver. If you just want to try out openTCS without an actual physical vehicle, the loopback driver (virtual vehicle) is the way to go.

**Reply by MinaTofiles** [reply]
Can't I use this example comm adapter with a simulator ?
I mean is there any way I could use this adapter without real physical vehicle ?

**Comment by swltr** [comment]
Yes, you could write a vehicle emulation and use that vehicle driver with the fake vehicle. (The "simulation" code already existing in the example integration project only fakes a few state messages. It does not fully emulate a vehicle that you could use to process transport orders.)

**Reply by MinaTofiles** [reply]
Thanks for your response, can I use loopback driver with some modifications to be in a real life project?

**Reply by swltr** [reply]
Technically, you can use the loopback driver as a template, of course. But the driver code in the example integration project may be better suited for that, as it actually implements a communication interface to the vehicle (which you can replace with your own code).

**Reply by MinaTofiles** [reply]
Thanks a lot for your help, after what you said and after digging into the source code that much I've understood that the example integration project is very similar to what I need. But I got stuck for a long time because I cannot connect the vehicle to the kernel. To simplify I started a  vehicle server with port 2000 but it does not respond with the correct format that opentcs needs. 
I am asking if there is any documentation I am missing or anything that could help me through this connection.
Also is there any vehicle code example provided by opentcs?
Thanks again for your help.

What i am getting:
No response to previous state request, yet - not sending another one.
[20230226-21:37:13-496] WARNING kernelExecutor       c.e.c.vehicle.ExampleCommAdapter.requestStateFromVehicle() : No response to previous state request, yet - not sending another one.
[20230226-21:37:13-996] WARNING kernelExecutor       c.e.c.vehicle.ExampleCommAdapter.requestStateFromVehicle() : No response to previous state request, yet - not sending another one.

**Reply by swltr** [reply]
You should actually replace the code responsible for communicating with your vehicle with your own code, i.e. implement the communication protocol that your vehicles speaks.

**Reply by MinaTofiles** [reply]
Hello, 
I've done what you said and now I can connect the vehicle with the kernel, but my vehicle is sending precise positions (x,y,z) and I cannot make a transport order unless it is on a specified point. I know that the vehicle must find a path located in the model so it can move on.
But I think I am missing something because am sure every real project will depend on either geolocation or precise position. I thought that I can make something like a dictionary that includes points and their respective (x,y,z), also the function that will retrieve the point will make an approximation to find something in the specified dictionary, I think this solution is too low for the quality. Is there any proposed solution for this?
Thank you for your fast response.

**Reply by swltr** [reply]
Precise positions / coordinates are optional for openTCS and only used for visualization. What is important is that the vehicle driver reports the logical position, i.e. the point name, to the kernel. So either the vehicle needs to reports its current/last point ID to the vehicle driver and the vehicle driver only forwards it to the kernel, or the vehicle driver needs to derive the current/last point ID from the coordinates and report that derived point ID to the kernel.
